In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

## Read Data

In [3]:
movies = pd.read_csv('../library/data/ml-latest-small/movies.csv')
ratings = pd.read_pickle('../library/data/ml-latest-small/ratings_concat.p')
genres = pd.read_pickle('../library/data/ml-latest-small/genres.p')

In [4]:
genres.sample()

(no genres listed)  Action  Adventure  Animation  Children  Comedy  \
movieId                                                                       
31260                     0       0          0          0         0       0   

         Crime  Documentary  Drama  Fantasy  Film-Noir  Horror  IMAX  Musical  \
movieId                                                                         
31260        0            0      1        0          0       0     0        0   

         Mystery  Romance  Sci-Fi  Thriller  War  Western  
movieId                                                    
31260          0        0       0         0    0        0

In [5]:
ratings.sample()

userId  movieId  rating            timestamp
63783     414     4822     2.0  2004-01-05 15:03:43

## Preprocessing

In [6]:
ratings = ratings.merge(genres, how='inner', left_on='movieId', right_index=True)

In [7]:
ratings.groupby('movieId').head(20)

userId  movieId  rating            timestamp  (no genres listed)  \
0            1        1     4.0  2000-07-30 18:45:03                   0   
516          5        1     4.0  1996-11-08 06:36:02                   0   
874          7        1     4.5  2005-01-25 06:52:26                   0   
1434        15        1     2.5  2017-11-13 12:59:30                   0   
1667        17        1     4.5  2011-05-18 05:28:03                   0   
1772        18        1     3.5  2016-02-11 16:56:56                   0   
2274        19        1     4.0  2000-08-08 03:33:57                   0   
3219        21        1     3.5  2014-08-09 21:14:38                   0   
4059        27        1     3.0  2000-07-04 04:34:22                   0   
4879        31        1     5.0  1996-12-13 08:43:36                   0   
4929        32        1     3.0  1997-02-23 22:15:19                   0   
5031        33        1     3.0  1999-10-11 13:10:44                   0   
5555        40        1     5.0  1996-05-14 07:29:19                   0   
6315        43        1     5.0  1996-11-26 07:39:43                   0   
6429        44        1     3.0  1997-07-18 18:51:00                   0   
6477        45        1     4.0  2000-02-21 21:56:22                   0   
6876        46        1     5.0  1996-06-14 21:31:46                   0   
7112        50        1     3.0  2017-12-25 21:41:56                   0   
7931        54        1     3.0  1996-04-23 08:15:30                   0   
8035        57        1     5.0  2000-08-09 04:40:31                   0   
1            1        3     4.0  2000-07-30 18:20:47                   0   
561          6        3     5.0  1996-10-17 12:11:36                   0   
2276        19        3     3.0  2000-08-08 04:07:16                   0   
4930        32        3     3.0  1997-02-23 22:16:12                   0   
5875        42        3     4.0  2001-07-27 08:04:05                   0   
6316        43        3     5.0  1996-11-26 07:46:45                   0   
6430        44        3     3.0  1997-07-18 18:51:50                   0   
7423        51        3     4.0  2009-01-02 21:45:36                   0   
8511        58        3     3.0  1996-11-11 13:36:37                   0   
9429        64        3     3.5  2006-10-22 12:21:08                   0   
...        ...      ...     ...                  ...                 ...   
100749     610   129313     4.0  2016-11-19 08:13:54                   0   
100751     610   130050     3.0  2017-05-03 22:21:51                   0   
100752     610   130052     3.5  2017-05-03 22:02:05                   0   
100754     610   130840     4.5  2017-05-03 21:57:01                   0   
100759     610   133832     3.5  2017-05-03 21:59:56                   0   
100765     610   135534     3.5  2017-05-03 21:56:04                   0   
100768     610   135803     3.5  2016-11-19 08:29:20                   0   
100769     610   135815     4.0  2016-11-19 08:29:48                   0   
100778     610   138610     1.5  2017-05-03 22:02:56                   0   
100779     610   138632     5.0  2016-11-19 08:50:20                   0   
100781     610   139511     3.0  2016-11-19 08:50:00                   0   
100783     610   139655     3.0  2017-05-03 21:57:19                   0   
100786     610   140267     4.0  2017-05-03 21:51:08                   0   
100787     610   141400     4.0  2016-11-19 08:29:30                   0   
100789     610   141799     3.5  2017-05-03 22:00:20                   0   
100790     610   142366     3.0  2016-11-19 08:12:26                   0   
100793     610   142598     4.0  2017-05-03 22:01:18                   0   
100797     610   145951     0.5  2017-05-03 22:03:39                   0   
100798     610   146309     4.0  2017-05-03 22:23:07                   0   
100799     610   147657     4.0  2016-11-19 08:30:10                   0   
100800     610   147662     3.0  2016-11-19 08:30:14    

In [8]:
ratings = ratings.replace(0, np.nan)
ratings.sample()

userId  movieId  rating            timestamp  (no genres listed)  \
43505     292      296     3.0  2001-05-17 01:23:18                 NaN   

       Action  Adventure  Animation  Children  Comedy   ...     Film-Noir  \
43505     NaN        NaN        NaN       NaN     1.0   ...           NaN   

       Horror  IMAX  Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  
43505     NaN   NaN      NaN      NaN      NaN     NaN       1.0  NaN      NaN  

[1 rows x 24 columns]

## train test split

In [9]:
from sklearn.model_selection import train_test_split
train, test = train_test_split(ratings, random_state=4444, test_size=.1)

In [10]:
print(train.shape)
print(test.shape)

(90763, 24)
(10085, 24)


## Item profile

In [11]:
genres.loc[4896]

(no genres listed)    0
Action                0
Adventure             1
Animation             0
Children              1
Comedy                0
Crime                 0
Documentary           0
Drama                 0
Fantasy               1
Film-Noir             0
Horror                0
IMAX                  0
Musical               0
Mystery               0
Romance               0
Sci-Fi                0
Thriller              0
War                   0
Western               0
Name: 4896, dtype: int64

## User profile

In [12]:
genre_cols = genres.columns
genre_cols

Index(['(no genres listed)', 'Action', 'Adventure', 'Animation', 'Children',
       'Comedy', 'Crime', 'Documentary', 'Drama', 'Fantasy', 'Film-Noir',
       'Horror', 'IMAX', 'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller',
       'War', 'Western'],
      dtype='object')

In [13]:
for cols in genre_cols:
    train[cols] =  train[cols] * train['rating']

In [14]:
train.head()
# one-hot 이었던 value들에 rating이 곱해진 것을 확인할 수 있음.

userId  movieId  rating            timestamp  (no genres listed)  \
81111     514      597     2.5  2018-08-11 01:01:16                 NaN   
5362       37      316     5.0  1996-10-13 20:44:02                 NaN   
68233     439     5952     4.0  2015-07-15 07:06:51                 NaN   
30350     212     8961     3.0  2018-04-08 19:48:40                 NaN   
36973     249    62081     3.5  2014-11-08 03:05:20                 NaN   

       Action  Adventure  Animation  Children  Comedy   ...     Film-Noir  \
81111     NaN        NaN        NaN       NaN     2.5   ...           NaN   
5362      5.0        5.0        NaN       NaN     NaN   ...           NaN   
68233     NaN        4.0        NaN       NaN     NaN   ...           NaN   
30350     3.0        3.0        3.0       3.0     3.0   ...           NaN   
36973     3.5        NaN        NaN       NaN     NaN   ...           NaN   

       Horror  IMAX  Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  
81111     NaN   NaN      NaN      NaN      2.5     NaN       NaN  NaN      NaN  
5362      NaN   NaN      NaN      NaN      NaN     5.0       NaN  NaN      NaN  
68233     NaN   NaN      NaN      NaN      NaN     NaN       NaN  NaN      NaN  
30350     NaN   NaN      NaN      NaN      NaN     NaN       NaN  NaN      NaN  
36973     NaN   3.5      NaN      NaN      NaN     NaN       3.5  NaN      NaN  

[5 rows x 24 columns]

In [15]:
train.groupby('userId')['Action'].mean()
# Action 장르에서 유저별 평점 평균

userId
1       4.345679
2       4.000000
3       3.461538
4       3.391304
5       3.111111
6       3.637931
7       3.283019
8       3.363636
9       3.000000
10      3.613636
11      3.450000
12      4.000000
13      4.375000
14      3.400000
15      3.140000
16      3.586957
17      4.236111
18      3.604972
19      2.705882
20      2.822222
21      3.463235
22      3.166667
23      3.540000
24      3.602041
25      4.892857
26      3.000000
27      3.789474
28      3.078212
29      3.960000
30      4.772727
          ...   
582     4.068966
583     3.000000
584     4.275862
585     4.277778
586     4.414286
587     4.142857
588     3.058824
589     3.750000
590     3.277523
591     3.846154
592     3.648649
593     2.962963
594     4.404110
595     4.000000
596     3.397661
597     3.577586
598     4.300000
599     2.729136
600     2.618644
601     4.390625
602     3.540541
603     3.152778
604     3.633333
605     3.232759
606     3.170455
607     3.734375
608     3.339921
609    

* User profile

In [16]:
user_profile = train.groupby('userId')[genre_cols].mean()
user_profile

(no genres listed)    Action  Adventure  Animation  Children  \
userId                                                                 
1                      NaN  4.345679   4.386667   4.703704  4.589744   
2                      NaN  4.000000   4.166667        NaN       NaN   
3                      NaN  3.461538   2.727273   0.500000  0.500000   
4                      NaN  3.391304   3.592593   4.000000  3.800000   
5                      NaN  3.111111   3.250000   4.333333  4.111111   
6                      NaN  3.637931   3.931818   4.166667  3.619048   
7                      NaN  3.283019   3.367347   3.392857  3.200000   
8                      NaN  3.363636   3.545455   5.000000  4.250000   
9                      NaN  3.000000   3.666667   4.000000  4.000000   
10                     NaN  3.613636   3.537037   3.846154  3.541667   
11                     NaN  3.450000   3.500000        NaN       NaN   
12                     NaN  4.000000   3.000000   3.000000       NaN   
13                     NaN  4.375000   4.200000        NaN       NaN   
14                     NaN  3.400000   3.444444        NaN  3.000000   
15                     NaN  3.140000   3.290000   2.954545  2.690476   
16                     NaN  3.586957   3.781250   3.950000  4.000000   
17                     NaN  4.236111   4.241935   4.312500  4.214286   
18                     NaN  3.604972   3.657480   3.729167  3.562500   
19                     NaN  2.705882   2.780142   2.902439  2.666667   
20                     NaN  2.822222   3.430380   4.258333  3.960000   
21                4.000000  3.463235   3.683544   3.220000  3.203704   
22                     NaN  3.166667   2.710526   2.600000  2.888889   
23                     NaN  3.540000   3.375000   3.777778  3.500000   
24                     NaN  3.602041   3.633333   4.250000  4.250000   
25                     NaN  4.892857   4.791667   4.666667  4.666667   
26                     NaN  3.000000   2.666667        NaN  3.000000   
27                     NaN  3.789474   3.740000   3.529412  3.261905   
28                     NaN  3.078212   3.178899   3.166667  2.500000   
29                     NaN  3.960000   4.062500   3.500000  4.250000   
30                     NaN  4.772727   4.812500   4.750000  5.000000   
...                    ...       ...        ...        ...       ...   
582                    NaN  4.068966   4.173913   3.833333  4.333333   
583                    NaN  3.000000   3.437500   3.388889  3.454545   
584                    NaN  4.275862   4.444444   4.800000  4.333333   
585                    NaN  4.277778   4.166667   4.000000       NaN   
586                    NaN  4.414286   4.480392   4.619048  4.521739   
587                    NaN  4.142857   4.120000   4.333333  3.875000   
588                    NaN  3.058824   2.800000   3.000000  3.000000   
589                    NaN  3.750000   4.000000        NaN  3.333333   
590                    NaN  3.277523   3.336207   3.833333  3.358333   
591                    NaN  3.846154   4.181818   3.666667  3.666667   
592                    NaN  3.648649   3.821429   3.833333  3.714286   
593                    NaN  2.962963   2.967742   3.142857  3.000000   
594                    NaN  4.404110   4.287879   4.600000  4.357143   
595                    NaN  4.000000   4.333333   4.500000  4.500000   
596               3.000000  3.397661   3.544521   4.045455  3.845238   
597                    NaN  3.577586   3.770270   3.857143  4.176471   
598                    NaN  4.300000   3.250000   1.000000  1.000000   
599               3.166667  2.729136   2.765509   2.959184  2.412338   
600                    NaN  2.618644   3.111511   3.232143  2.990385   
601               4.750000  4.390625   4.448718   4.571429  4.555556   
602                    NaN  3.540541   3.230769   4.000000  3.000000   
603                    NaN  3.152778   3.021505   1.666667  3.066667   
604                    NaN  3.633333   3.666667 

* My user profile

In [19]:
user_profile.loc[1000]

(no genres listed)         NaN
Action                3.333333
Adventure             2.000000
Animation             2.500000
Children              3.000000
Comedy                3.500000
Crime                 5.000000
Documentary                NaN
Drama                 4.800000
Fantasy               3.500000
Film-Noir                  NaN
Horror                     NaN
IMAX                  4.000000
Musical               2.000000
Mystery               4.750000
Romance               4.000000
Sci-Fi                3.500000
Thriller              4.750000
War                        NaN
Western                    NaN
Name: 1000, dtype: float64

## Predict one sample

In [20]:
sample = test.loc[13852]
sample

userId                                 89
movieId                             88069
rating                                  4
timestamp             2018-03-07 07:59:16
(no genres listed)                    NaN
Action                                NaN
Adventure                             NaN
Animation                             NaN
Children                              NaN
Comedy                                  1
Crime                                   1
Documentary                           NaN
Drama                                 NaN
Fantasy                               NaN
Film-Noir                             NaN
Horror                                NaN
IMAX                                  NaN
Musical                               NaN
Mystery                               NaN
Romance                               NaN
Sci-Fi                                NaN
Thriller                              NaN
War                                   NaN
Western                           

In [21]:
sample_user = sample['userId']

In [22]:
sample_user_profile = user_profile.loc[sample_user]
sample_user_profile

(no genres listed)    3.000000
Action                3.367347
Adventure             3.563291
Animation             4.063291
Children              3.918478
Comedy                3.465197
Crime                 3.171053
Documentary           4.500000
Drama                 3.430851
Fantasy               3.415094
Film-Noir                  NaN
Horror                4.242424
IMAX                  4.000000
Musical               3.138889
Mystery               3.700000
Romance               3.122549
Sci-Fi                3.483333
Thriller              3.738095
War                   3.272727
Western               4.500000
Name: 89, dtype: float64

* sample user profile로 sample 대한 예측

In [23]:
print(sample['movieId'])
sample[genre_cols]

88069


(no genres listed)    NaN
Action                NaN
Adventure             NaN
Animation             NaN
Children              NaN
Comedy                  1
Crime                   1
Documentary           NaN
Drama                 NaN
Fantasy               NaN
Film-Noir             NaN
Horror                NaN
IMAX                  NaN
Musical               NaN
Mystery               NaN
Romance               NaN
Sci-Fi                NaN
Thriller              NaN
War                   NaN
Western               NaN
Name: 13852, dtype: object

* comedy 와 crime 하나씩 있으니, 각각을 sample_user_profile에 있는 장르 평점과 곱해주고, mean

In [24]:
(sample[genre_cols] * sample_user_profile).mean()

3.3181249236781047

* sample에 대해, we predict that the user will socre 3.318..

## Predict all testsets

In [25]:
from tqdm import tqdm_notebook

In [26]:
predict = []
for idx, row in tqdm_notebook(test.iterrows()):
    user = row['userId']
#     item profile * user profile 
    predict.append((row[genre_cols] * user_profile.loc[user]).mean())

In [36]:
test['predict'] = predict
test.head(30)

userId  movieId  rating            timestamp  (no genres listed)  \
55414     368      161     4.0  2000-10-11 14:34:26                 NaN   
95358     600     1682     3.5  2009-03-22 09:08:46                 NaN   
81635     517     1270     5.0  2017-02-24 17:18:10                 NaN   
91184     590    48738     4.0  2009-11-17 03:44:12                 NaN   
3752       22    53519     0.5  2010-03-16 08:12:17                 NaN   
35991     246      910     4.5  2012-11-28 18:22:45                 NaN   
20425     135     1358     5.0  2001-12-30 05:45:44                 NaN   
40413     274    55118     4.0  2008-02-20 02:06:53                 NaN   
35276     238     3101     4.0  2001-12-09 20:50:50                 NaN   
57756     380    52950     2.0  2017-05-08 18:29:06                 NaN   
26471     182     2359     3.0  2003-06-05 03:36:15                 NaN   
59141     385     1291     5.0  1997-05-30 20:55:17                 NaN   
74226     474     4129     1.0  2003-03-07 16:41:10                 NaN   
17338     111     2572     4.0  2018-01-17 01:48:19                 NaN   
71645     462     1283     4.5  2005-08-13 00:11:39                 NaN   
63256     414     2881     2.0  2000-06-19 17:44:18                 NaN   
82016     520     1198     5.0  2012-01-15 06:30:42                 NaN   
23610     160     3826     1.0  2001-08-22 20:53:05                 NaN   
68008     438     5902     4.0  2005-01-13 22:06:13                 NaN   
85009     552     2502     5.0  2005-03-22 06:33:10                 NaN   
56694     377     1274     5.0  2012-06-22 05:35:15                 NaN   
46537     305   109850     3.5  2016-04-11 09:15:58                 NaN   
95375     600     1923     3.0  2009-03-22 09:08:41                 NaN   
64387     414    30848     3.5  2008-07-15 18:46:33                 NaN   
47470     307     7361     5.0  2007-08-03 17:28:37                 NaN   
33392     226     5110     5.0  2004-09-29 01:22:40                 NaN   
71712     462     2693     4.5  2008-01-27 14:57:57                 NaN   
25853     178     4002     4.0  2006-11-16 10:40:20                 NaN   
46897     307     1089     4.0  2007-08-03 17:17:45                 NaN   
77708     483     8644     4.0  2007-05-05 09:48:39                 NaN   

       Action  Adventure  Animation  Children  Comedy    ...     Horror  IMAX  \
55414     NaN        NaN        NaN       NaN     NaN    ...        NaN   NaN   
95358     NaN        NaN        NaN       NaN     1.0    ...        NaN   NaN   
81635     NaN        1.0        NaN       NaN     1.0    ...        NaN   NaN   
91184     NaN        NaN        NaN       NaN     NaN    ...        NaN   NaN   
3752      1.0        1.0        NaN       NaN     NaN    ...        1.0   NaN   
35991     NaN        NaN        NaN       NaN     1.0    ...        NaN   NaN   
20425     NaN        NaN        NaN       NaN     NaN    ...        NaN   NaN   
40413     NaN        NaN        NaN       NaN     NaN    ...        NaN   NaN   
35276     NaN        NaN        NaN       NaN     NaN    ...        NaN   NaN   
57756     1.0        NaN        NaN       NaN     NaN    ...        1.0   NaN   
26471     NaN        NaN        NaN       NaN     1.0    ...        NaN   NaN   
59141     1.0        1.0        NaN       NaN     NaN    ...        NaN   NaN   
74226     NaN        NaN        NaN       NaN     NaN    ...        NaN   NaN   
17338     NaN        NaN        NaN       NaN     1.0    ...        NaN   NaN   
71645     NaN        NaN        NaN       NaN     NaN    ...        NaN   NaN   
63256     1.0        NaN        NaN       NaN     NaN    ...        NaN   NaN   
82016     1.0        1.0        NaN       NaN     NaN    ...        NaN   NaN   
23610     NaN        NaN        NaN       NaN     NaN    ...        1.0   NaN   
68008     NaN        NaN        NaN       NaN     1.0    ...        NaN   NaN   
85009     NaN        NaN        NaN       NaN     1.0    ... 

* null 값이 있는 case들이 있다.

In [30]:
test[test['predict'].isnull()]

userId  movieId  rating            timestamp  (no genres listed)  \
11668      70      246     4.5  2012-12-11 04:06:33                 NaN   
81466     514   182727     2.0  2018-09-08 04:30:35                 1.0   
79767     496      904     5.0  2014-11-05 05:50:05                 NaN   
84755     548    80862     3.5  2017-05-15 20:30:32                 NaN   
79902     499    38304     4.5  2010-11-17 02:17:40                 NaN   
79793     496   114713     2.5  2014-11-09 08:17:27                 NaN   
18794     121      246     3.0  1996-11-10 20:06:14                 NaN   
61119     398    82459     4.0  2011-07-21 00:24:35                 NaN   
37495     252   122896     3.0  2017-06-24 06:01:10                 1.0   
90221     586   122896     5.0  2018-06-25 04:06:27                 1.0   
37887     256   121171     4.0  2015-11-08 16:18:42                 NaN   
0        1000   168250     5.0  2020-06-13 22:09:34                 NaN   
1090        9     2877     3.0  2003-02-07 22:33:41                 NaN   
11717      70     5669     4.5  2012-12-11 04:07:01                 NaN   
81449     514   149330     4.0  2018-09-03 03:06:45                 1.0   
80167     503    94130     3.5  2012-04-23 22:16:16                 NaN   

       Action  Adventure  Animation  Children  Comedy   ...     Horror  IMAX  \
11668     NaN        NaN        NaN       NaN     NaN   ...        NaN   NaN   
81466     NaN        NaN        NaN       NaN     NaN   ...        NaN   NaN   
79767     NaN        NaN        NaN       NaN     NaN   ...        NaN   NaN   
84755     NaN        NaN        NaN       NaN     NaN   ...        NaN   NaN   
79902     NaN        NaN        NaN       NaN     NaN   ...        NaN   NaN   
79793     NaN        NaN        NaN       NaN     NaN   ...        1.0   NaN   
18794     NaN        NaN        NaN       NaN     NaN   ...        NaN   NaN   
61119     NaN        NaN        NaN       NaN     NaN   ...        NaN   NaN   
37495     NaN        NaN        NaN       NaN     NaN   ...        NaN   NaN   
90221     NaN        NaN        NaN       NaN     NaN   ...        NaN   NaN   
37887     NaN        NaN        NaN       NaN     NaN   ...        NaN   NaN   
0         NaN        NaN        NaN       NaN     NaN   ...        1.0   NaN   
1090      NaN        NaN        NaN       NaN     NaN   ...        NaN   NaN   
11717     NaN        NaN        NaN       NaN     NaN   ...        NaN   NaN   
81449     NaN        NaN        NaN       NaN     NaN   ...        NaN   NaN   
80167     NaN        NaN        NaN       NaN     NaN   ...        NaN   NaN   

       Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  predict  
11668      NaN      NaN      NaN     NaN       NaN  NaN      NaN      NaN  
81466      NaN      NaN      NaN     NaN       NaN  NaN      NaN      NaN  
79767      NaN      1.0      NaN     NaN       1.0  NaN      NaN      NaN  
84755      NaN      1.0      NaN     NaN       NaN  NaN      NaN      NaN  
79902      NaN      NaN      NaN     NaN       NaN  NaN      NaN      NaN  
79793      NaN      NaN      NaN     NaN       NaN  NaN      NaN      NaN  
18794      NaN      NaN      NaN     NaN       NaN  NaN      NaN      NaN  
61119      NaN      NaN      NaN     NaN       NaN  NaN      1.0      NaN  
37495      NaN      NaN      NaN     NaN       NaN  NaN      NaN      NaN  
90221      NaN      NaN      NaN     NaN       NaN  NaN      NaN      NaN  
37887      NaN      NaN      NaN     NaN       NaN  NaN      NaN      NaN  
0          NaN      NaN      NaN     NaN       NaN  NaN      NaN      NaN  
1090       1.0      NaN      NaN     NaN       NaN  NaN      NaN      NaN  
11717      NaN      NaN      NaN     NaN       NaN  NaN      NaN      NaN  
81449      NaN      NaN      NaN     NaN       NaN  NaN      NaN      NaN  
80167      NaN      NaN      NaN     NaN       NaN  NaN      NaN      NaN  

[16 rows x 25 columns]

* global mean으로 채우자!

In [39]:
test.loc[test['predict'].isnull(), 'predict'] = train['rating'].mean()

In [40]:
test[test['predict'].isnull()]

Empty DataFrame
Columns: [userId, movieId, rating, timestamp, (no genres listed), Action, Adventure, Animation, Children, Comedy, Crime, Documentary, Drama, Fantasy, Film-Noir, Horror, IMAX, Musical, Mystery, Romance, Sci-Fi, Thriller, War, Western, predict]
Index: []

[0 rows x 25 columns]

## Model validation

In [41]:
from sklearn.metrics import mean_squared_error

In [43]:
mse = mean_squared_error(test['rating'], test['predict'])
rmse = np.sqrt(mse)
rmse

0.918052219583595